File: src/main/java/sx/concurrent/AutoResetEvent.java
package sx.concurrent;

import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

/**
 * Automatic reset mSemaphore
 * @author masc
 */
public class AutoResetEvent implements Event {
    private final Semaphore mSemaphore;

    public AutoResetEvent(boolean signalled) {
        mSemaphore = new Semaphore(signalled ? 1 : 0);
    }

    public void set() {
        synchronized (mSemaphore) {
            if (mSemaphore.availablePermits() == 0)
                mSemaphore.release();
        }
    }

    public void reset() {
        mSemaphore.drainPermits();
    }

    public void waitOne() throws InterruptedException {
        mSemaphore.acquire();
    }

    public boolean waitOne(int timeout, TimeUnit unit) throws InterruptedException {
        return mSemaphore.tryAcquire(timeout, unit);
    }

    public boolean isSignalled() {
        return mSemaphore.availablePermits() > 0;
    }
}

File: src/main/java/sx/concurrent/AutoResetEvent.kt
package sx.concurrent

import java.util.concurrent.Semaphore
import java.util.concurrent.TimeUnit

/**
 * Automatic reset mSemaphore
 * @author masc
 */
class AutoResetEvent(signalled: Boolean) : Event {
    private val semaphore: Semaphore

    init {
        semaphore = Semaphore(if (signalled) 1 else 0)
    }

    override fun set() {
        synchronized (semaphore) {
            if (semaphore.availablePermits() == 0)
                semaphore.release()
        }
    }

    override fun reset() {
        semaphore.drainPermits()
    }

    @Throws(InterruptedException::class)
    override fun waitOne() {
        semaphore.acquire()
    }

    @Throws(InterruptedException::class)
    override fun waitOne(timeout: Int, unit: TimeUnit): Boolean {
        return semaphore.tryAcquire(timeout.toLong(), unit)
    }

    override val isSignalled: Boolean
        get() = semaphore.availablePermits() > 0
}